src/components/VGrid/VGrid.tsx

```diff
@@ -0,0 +1,298 @@
+      : 'repeat(1, 1fr)',
+  } as const;
+};
+
+const prerenderRowsLength = 1;
+
+export class VGrid<T, K extends keyof T> extends React.Component<Props<T, K>, State> {
+  static contextType = InitialHashContext;
+
+  context!: InitialHashContextValue;
+  containerRef = React.createRef<HTMLDivElement>();
+  intersectionObserver?: IntersectionObserver;
+  resizeObserver?: ResizeObserver;
+  gridProperties: { media: MediaQueryList; prop: GridStyleProperty }[];
+  previousContainerWidth?: number;
+
+  state: State = {
+    isIntersecting: false,
+    offsetIndex: 0,
+    repeatLength: 1,
+    gridStyle: {
+      padding: 0,
+      margin: 0,
+      listStyle: 'none',
+      display: 'grid',
+      position: 'absolute',
+      left: 0,
+      right: 0,
+      gridGap: -1,
+      gridTemplateColumns: '',
+    },
+    containerHeight: 0,
+    visibleItemsLength: 20,
+  };
+
+  constructor(props: Props<T, K>) {
+    super(props);
+
+    this.handleOnScroll = this.handleOnScroll.bind(this);
+    this.handleOnHashChange = this.handleOnHashChange.bind(this);
+
+    // Create pseudo CSS obj to emulate the media query such as:
+    //
+    // ```css
+    // @media screen and (min-width: 320px) {
+    //   ...
+    // }
+    // ```
+    // The created pair MediaQueryList instance will be used to determine grid style properties
+    //
+    if (!Array.isArray(props.gridOptions)) {
+      this.gridProperties = [
+        {
+          media: matchMedia('all'),
+          prop: props.gridOptions,
+        },
+      ];
+    } else {
+      this.gridProperties = props.gridOptions
+        .slice()
+        .reverse()
+        .map(({ media, ...rest }) => ({
+          media: matchMedia(media),
+          prop: rest,
+        }));
+    }
+  }
+
+  componentDidMount() {
+    const containerElement = this.containerRef.current;
+    if (!containerElement) return;
+
+    // Using IntersectionObserver to prevent scrolling calculation when the entire container is out of viewport.
+    this.intersectionObserver = new IntersectionObserver(
+      (entries) => {
+        if (entries.length !== 1) return;
+        const entry = entries[0];
+        const { isIntersecting } = entry;
+        this.setState({ isIntersecting });
+      },
+      {
+        threshold: 0,
+      },
+    );
+    this.intersectionObserver.observe(containerElement);
+
+    // Using ResizeObserver because we should update some grid properties(e.g. the num of cells per 1 row) when the container element is resized.
+    this.resizeObserver = new ResizeObserver((entries) => {
+      if (entries.length !== 1) return;
+      const entry = entries[0];
+      const {
+        contentRect: { width },
+      } = entry;
+      if (this.previousContainerWidth !== width) {
+        this.updateContainerState(width);
+      }
+      this.previousContainerWidth = width;
+    });
+    this.resizeObserver.observe(containerElement);
+
+    document.addEventListener('scroll', this.handleOnScroll);
```

Comment:
```suggestion
    document.addEventListener('scroll', this.handleOnScroll, { passive: true });
```

I want to enable `passive` for scroll event that don't need to call `preventDefault` :dog: